src/main/java/org/openjdk/jextract/impl/NameMangler.java

```diff
@@ -318,30 +320,46 @@ private static boolean isRestrictedTypeName(String name) {
         };
     }
 
+
+    // well known java type names that are potentially used
+    // as unqualified name in the jextract generated code.
+    private static final Set<String> WELL_KNOWN_JAVA_TYPES = ConcurrentHashMap.newKeySet();
+    static {
+        WELL_KNOWN_JAVA_TYPES.add("ByteOrder");
+        WELL_KNOWN_JAVA_TYPES.add("MethodHandle");
+        WELL_KNOWN_JAVA_TYPES.add("VarHandle");
+    }
+
     private static boolean isJavaTypeName(String name) {
         // rule out common case of lower-case starting identifier.
         if (name.length() == 0 || Character.isLowerCase(name.charAt(0))) {
             return false;
         }
 
         // Java types that are used unqualified in the generated code
-        return switch (name) {
-            case "String", "Struct", "MethodHandle",
-                "VarHandle", "ByteOrder",
-                "FunctionDescriptor", "LibraryLookup",
-                "MemoryLayout",
-                "Arena", "NativeArena", "MemorySegment", "ValueLayout"
-                    -> true;
-            default -> {
-                try {
-                    // java.lang package is auto-imported. Check if we've clash
-                    // of the given identifier with any of the java.lang.* types.
-                    Class.forName("java.lang." + name);
-                    yield true;
-                } catch (Exception ignored) {
-                }
-                yield false;
+        if (WELL_KNOWN_JAVA_TYPES.contains(name)) {
+            return true;
+        }
+
+        // may be it is not cached as well-known java type yet.
+        if (checkTypeInPackage("java.lang", name)) {
+            return true;
+        } else if (checkTypeInPackage("java.lang.foreign", name)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private static boolean checkTypeInPackage(String pkgName, String typeName) {
+        try {
+            int mods = Class.forName(pkgName + "." + typeName).getModifiers();
+            if (Modifier.isPublic(mods)) {
+                WELL_KNOWN_JAVA_TYPES.add(typeName);
+                return true;
             }
-        };
+        } catch (Exception ignored) {
```

Comment:
Should we report something when using the verbose debug option?

Reply:
Hmm. Not sure. May be bit noisy I think as exceptions will be thrown for many uppercase starting C identifiers (enum constants, #defines)